autoHeightImage.js

```diff
@@ -3,102 +3,47 @@
  * @author vivaxy
  */
 
-import React, { PureComponent } from 'react';
+import React, { useEffect, useState } from 'react';
 import ImagePolyfill from './imagePolyfill';
-import { StyleSheet, Image } from 'react-native';
+import { Image } from 'react-native';
 import PropTypes from 'prop-types';
 
-import { getImageSizeFitWidth, getImageSizeFitWidthFromCache } from './cache';
+import { getImageSizeFitWidth } from './cache';
 import { NOOP, DEFAULT_HEIGHT } from './helpers';
 
 // remove `resizeMode` props from `Image.propTypes`
 const { resizeMode, ...ImagePropTypes } = Image.propTypes;
 
-export default class AutoHeightImage extends PureComponent {
-  static propTypes = {
-    ...ImagePropTypes,
-    width: PropTypes.number.isRequired,
-    onHeightChange: PropTypes.func
-  };
-
-  static defaultProps = {
-    onHeightChange: NOOP
-  };
-
-  constructor(props) {
-    super(props);
-    this.setInitialImageHeight();
-  }
-
-  updateSequence = 0;
-
-  async componentDidMount() {
-    this.hasMounted = true;
-    await this.updateImageHeight(this.props);
-  }
-
-  async componentDidUpdate() {
-    await this.updateImageHeight(this.props);
-  }
-
-  componentWillUnmount() {
-    this.hasMounted = false;
-    // clear memory usage
-    this.updateSequence = null;
-  }
-
-  setInitialImageHeight() {
-    const { source, width, onHeightChange } = this.props;
-    const { height = DEFAULT_HEIGHT } = getImageSizeFitWidthFromCache(
-      source,
-      width
-    );
-    this.state = { height };
-    this.styles = StyleSheet.create({ image: { width, height } });
-    onHeightChange(height);
-  }
-
-  async updateImageHeight(props) {
-    if (
-      this.state.height === DEFAULT_HEIGHT ||
-      this.props.width !== props.width ||
-      this.props.source !== props.source
-    ) {
-      // image height could not be `0`
-      const { source, width, onHeightChange } = props;
-      try {
-        const updateSequence = ++this.updateSequence;
-        const { height } = await getImageSizeFitWidth(source, width);
-        if (updateSequence !== this.updateSequence) {
-          return;
-        }
+function AutoHeightImage(props) {
+  const { onHeightChange, source, width, style, ...rest } = props;
+  const [height, setHeight] = useState(DEFAULT_HEIGHT);
+
+  useEffect(() => {
```

Comment:
```js
useEffect(() => {
  (async () => {
    // ...
  })();
}, [source, onHeightChange, width]);
```
Better be:
```js
useEffect(async () => {
  // ...
}, [source, onHeightChange, width]);
```

Reply:
This definitely looks better, but not correct. `useEffect` callbacks must not return anything besides a cleanup function, and your suggestion returns a Promise. That would result in this warning:
![image](https://user-images.githubusercontent.com/12103433/72548478-c31fa880-385c-11ea-8493-30503d810ccc.png)

More info: https://medium.com/javascript-in-plain-english/how-to-use-async-function-in-react-hook-useeffect-typescript-js-6204a788a435

Reply:
> This definitely looks better, but not correct. `useEffect` callbacks must not return anything besides a cleanup function, and your suggestion returns a Promise. That would result in this warning:
> ![image](https://user-images.githubusercontent.com/12103433/72548478-c31fa880-385c-11ea-8493-30503d810ccc.png)
> 
> More info: [medium.com/javascript-in-plain-english/how-to-use-async-function-in-react-hook-useeffect-typescript-js-6204a788a435](https://medium.com/javascript-in-plain-english/how-to-use-async-function-in-react-hook-useeffect-typescript-js-6204a788a435)

Thanks for the information!